### PR TITLE
Add focus requests

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Depends on https://github.com/benashford/rs-es/pull/133

### CODE REVIEW

File: libs/bragi/src/query.rs [hunk 1]
```diff
@@ -203,20 +234,14 @@ fn build_query<'a>(
...
-                    .with_missing(0.)
-                    .build(),
-            )
-            .with_boost_mode(BoostMode::Replace)
-            .build(),
+    let mut importance_queries = match (coord, focus) {
+        (Some(ref c), None) => vec![build_proximity_with_boost(c, 100.)],
+        (None, Some(ref c)) => vec![build_with_weight(0.5), build_focus_with_boost(c, 0.4)],
+        (None, None) => vec![build_with_weight(None)],
+        (Some(_), Some(_)) => panic!("you cannot pass both coord and focus positions at the same time"),
```
Comment (antoine-de) [inline libs/bragi/src/query.rs]: we don't want to panic in the API, we want to return an error

File: libs/bragi/src/routes/autocomplete.rs [hunk 2]
```diff
@@ -77,6 +77,8 @@ pub struct Params {
     #[serde(default, rename = "poi_type")]
     poi_types: Vec<PoiType>,
     lang: Option<String>,
+    focus_lat: Option<f64>,
```
Comment (antoine-de) [inline libs/bragi/src/routes/autocomplete.rs]: you should add a comment explaining the use of this parameter

**Comment by GuillaumeGomez** [comment]
`coord` excludes too far places whereas `focus` boosts the close ones.

File: libs/bragi/src/query.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply libs/bragi/src/query.rs]: Normally, at this point, we already checked that we didn't have both parameters given at the same time. I'll still return an error though (and therefore update the function signature).

**Comment by antoine-de** [comment]
I don't feel the need of 2 parameters here.

The current `coord` behavior is indeed not perfect, but I'll be more in favor of changing the `coord` behavior instead of adding a new param

**Comment by GuillaumeGomez** [comment]
We did it like this specifically to avoid breaking things on your side. If you're fine with it then I'll definitely update coord instead. :)

File: libs/bragi/src/query.rs [hunk 3]
```diff
@@ -130,19 +113,49 @@ fn build_coverage_condition(pt_datasets: &[&str]) -> Query {
...
     shape: Option<Geometry>,
     pt_datasets: &[&str],
     all_data: bool,
     langs: &'a [&'a str],
     zone_types: &[&str],
     poi_types: &[&str],
-) -> Query {
-    use rs_es::query::functions::Function;
-
+) -> Result<Query, String> {
```
Comment (antoine-de) [inline libs/bragi/src/query.rs]: why do you need this Result ?

Comment (GuillaumeGomez) [reply libs/bragi/src/query.rs]: I don't need it anymore, good catch!

**Comment by kinnou02** [comment]
> We did it like this specifically to avoid breaking things on your side. If you're fine with it then I'll definitely update coord instead. :)

To be clear: We are totally ok to change the behavior of `coord`

File: libs/bragi/src/query.rs [hunk 4]
```diff
@@ -203,20 +217,16 @@ fn build_query<'a>(
...
-                Function::build_field_value_factor("weight")
-                    .with_factor(0.15)
-                    .with_missing(0.)
-                    .build(),
-            )
-            .with_boost_mode(BoostMode::Replace)
-            .build(),
+    let mut importance_queries = if let Some(ref coord) = coord {
+        vec![
+            build_with_weight(0.5),
```
Comment (kinnou02) [inline libs/bragi/src/query.rs]: :+1:

File: libs/bragi/src/query.rs [hunk 5]
```diff
@@ -130,19 +113,50 @@ fn build_coverage_condition(pt_datasets: &[&str]) -> Query {
...
+/// distance to `coord`.
+fn build_proximity_with_boost(coord: &Coord, weight: f64) -> Query {
+    Query::build_function_score()
+        .with_functions(vec![
+            DecayOptions::new(
+                rs_u::Location::LatLon(coord.lat(), coord.lon()),
+                rs_u::Distance::new(1000f64, rs_u::DistanceUnit::Kilometer),
+            )
+            .with_offset(rs_u::Distance::new(50f64, rs_u::DistanceUnit::Kilometer))
+            .with_scale(Scale::Distance(rs_u::Distance::new(200f64, rs_u::DistanceUnit::Kilometer)))
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply libs/bragi/src/query.rs]: Good catch! I'll update it.

File: libs/bragi/src/query.rs [hunk 6]
```diff
@@ -352,7 +361,10 @@ fn query(
         langs,
         zone_types,
         poi_types,
-    );
+    ) {
```
Comment (riendegris) [inline libs/bragi/src/query.rs]: Ce n'est pas l'equivalent d'un '?' ?

Comment (GuillaumeGomez) [reply libs/bragi/src/query.rs]: Au moment où je l'ai écrit non car il y avait besoin de changer le type de `Result`.

Comment (riendegris) [reply libs/bragi/src/query.rs]: Si tu n'as pas besoin de changer le Result, est-ce possible de remettre le '?'. C'est plus consistent avec le reste du code, et rust en general. Thx

Comment (GuillaumeGomez) [reply libs/bragi/src/query.rs]: Done.

**Comment by GuillaumeGomez** [comment]
Current results:

```
                region                 | category | failed | total | ratio |    duration    
---------------------------------------|----------|--------|-------|-------|----------------
 test_admin_capital_cities_i18n_en.csv |  fuzzy   |   _    |   _   |  0%   | 0:00:02.130000 
 test_admin_capital_cities_i18n_en.csv |  admins  |  265   |  486  |  45%  | 0:00:05.620000 
 test_admin_capital_cities_i18n_en.csv |   poi    |   _    |   _   |  0%   | 0:00:02.130000 
 test_admin_capital_cities_i18n_en.csv |  autre   |   _    |   _   |  0%   | 0:00:02.180000 
         test_admin_cities.csv         |  fuzzy   |   _    |   _   |  0%   | 0:00:02.290000 
         test_admin_cities.csv         |  admins  |  390   |  940  |  59%  | 0:00:06.350000 
         test_admin_cities.csv         |   poi    |   _    |   _   |  0%   | 0:00:02.390000 
         test_admin_cities.csv         |  autre   |   _    |   _   |  0%   | 0:00:02.430000 
     test_admin_cities_i18n_en.csv     |  fuzzy   |   _    |   _   |  0%   | 0:00:02.330000 
     test_admin_cities_i18n_en.csv     |  admins  |  385   |  940  |  59%  | 0:00:06.270000 
     test_admin_cities_i18n_en.csv     |   poi    |   _    |   _   |  0%   | 0:00:02.440000 
     test_admin_cities_i18n_en.csv     |  autre   |   _    |   _   |  0%   | 0:00:02.320000 
     test_admin_cities_i18n_fr.csv     |  fuzzy   |   _    |   _   |  0%   | 0:00:02.320000 
     test_admin_cities_i18n_fr.csv     |  admins  |  401   |  940  |  57%  | 0:00:06.230000 
     test_admin_cities_i18n_fr.csv     |   poi    |   _    |   _   |  0%   | 0:00:02.340000 
     test_admin_cities_i18n_fr.csv     |  autre   |   _    |   _   |  0%   | 0:00:02.410000 
                   _                   |  fuzzy   |  258   |  993  |  74%  | 0:00:29.450000 
                   _                   |  admins  |  1965  | 4168  |  53%  | 0:01:02.700000 
                   _                   |   poi    |  2918  | 9521  |  69%  | 0:03:04.140000 
                   _                   |  autre   |  1999  | 11508 |  83%  | 0:02:27.310000
```

**Comment by kinnou02** [comment]
This have evolved a bit outside the context of the PR

**Comment by GuillaumeGomez** [comment]
No, we're still fully inside it. :)

**Comment by GuillaumeGomez** [comment]
The full details:

The elasticsearch mappings changes are because the current ngram isn't work (I invite you to change the ngram conditions to observe that there are no differences in the results of a same query). We discovered it while trying to improve performances (more than 1 second for a query was way too much).

To come back on the ngram analyzer not working: it's a known bug in elasticsearch and still exists to this day. You need to define the analyzer yourself to overcome it, which explains the change.

About the minimum_should_match conditions now (which is `1<-1 4<-3 20<15%`). From 1 to 4 ngram, we allow one mistake, from 5 to 20, we allow 3 and from 21 you only need to match to 15%.

**Comment by GuillaumeGomez** [comment]
Last version:

```
                region                 | category | failed | total | ratio |    duration    
---------------------------------------|----------|--------|-------|-------|----------------
 test_admin_capital_cities_i18n_en.csv |  fuzzy   |   _    |   _   |  0%   | 0:00:02.030000 
 test_admin_capital_cities_i18n_en.csv |  admins  |  265   |  486  |  45%  | 0:00:05.940000 
 test_admin_capital_cities_i18n_en.csv |   poi    |   _    |   _   |  0%   | 0:00:01.970000 
 test_admin_capital_cities_i18n_en.csv |  autre   |   _    |   _   |  0%   | 0:00:01.890000 
         test_admin_cities.csv         |  fuzzy   |   _    |   _   |  0%   | 0:00:02.380000 
         test_admin_cities.csv         |  admins  |  390   |  940  |  59%  | 0:00:06.830000 
         test_admin_cities.csv         |   poi    |   _    |   _   |  0%   | 0:00:02.880000 
         test_admin_cities.csv         |  autre   |   _    |   _   |  0%   | 0:00:02.190000 
     test_admin_cities_i18n_en.csv     |  fuzzy   |   _    |   _   |  0%   | 0:00:02.180000 
     test_admin_cities_i18n_en.csv     |  admins  |  386   |  940  |  59%  | 0:00:06.440000 
     test_admin_cities_i18n_en.csv     |   poi    |   _    |   _   |  0%   | 0:00:02.160000 
     test_admin_cities_i18n_en.csv     |  autre   |   _    |   _   |  0%   | 0:00:02.350000 
     test_admin_cities_i18n_fr.csv     |  fuzzy   |   _    |   _   |  0%   | 0:00:02.670000 
     test_admin_cities_i18n_fr.csv     |  admins  |  401   |  940  |  57%  | 0:00:06.930000 
     test_admin_cities_i18n_fr.csv     |   poi    |   _    |   _   |  0%   | 0:00:02.360000 
     test_admin_cities_i18n_fr.csv     |  autre   |   _    |   _   |  0%   | 0:00:02.370000 
                   _                   |  fuzzy   |  249   |  993  |  75%  | 0:00:28.300000 
                   _                   |  admins  |  1961  | 4168  |  53%  | 0:00:56.170000 
                   _                   |   poi    |  3307  | 9521  |  65%  | 0:03:20.100000 
                   _                   |  autre   |  2044  | 11508 |  82%  | 0:02:37.910000
```

**Comment by GuillaumeGomez** [comment]
Should be ready now!

File: json/addr_settings.json [hunk 7]
```diff
@@ -38,10 +38,16 @@
                     "char_filter" : [ ]
                 },
                 "ngram": {
-                    "type": "custom",
-                    "tokenizer": "standard",
-                    "filter": [ "lowercase", "asciifolding", "ngram_filter" ],
-                    "char_filter" : [ ]
+                    "tokenizer": "my_ngram_tokenizer",
```
Comment (kinnou02) [inline json/addr_settings.json]: you didn't update `ngram_with_synonyms` is that what you want?

Comment (GuillaumeGomez) [reply json/addr_settings.json]: Yes.

**Comment by GuillaumeGomez** [comment]
Explanations for the regressions: we now have an offset which makes all items in this area having the "weight" in the results. A solution for this would be to check for the X first results that what we're looking for is indeed in those results.

**Pull request merged**